### PR TITLE
Tensorflow Object Detection Pipeline

### PR DESCRIPTION
**Pull request by Grymessa**
This is now a feature complete, working pipeline for processing data and training new neural networks for object detection. There are a few quality of life adjustments and general clean up that still needs to be done with it however it is now in a working order and is fairly well documented. 

General TODO list: 

- [x] Generate a single script that can be run through rosrun and eliminate the shell script in data processing.
- [x] Have a file on the repo with premade labelmaps for easy selection 
- [ ] Better document and explain exactly what Tensorflow itself, the main training script, is doing.

### CODE REVIEW

**Review by kev-the-dev — COMMENTED** [review]
I'll give a closer review once the WIP tag is removed, but a summary of our conversation:

* Make root of this directory a catkin package
* Minimize file structure complexity / files. Please don't commit anything that is no longer used, including unused files, functions, or dead code
* Minimize any "hard coded" assumptions, such as what robots this will be used for. Things like what classes to train on and what model to use should be configurable.
* Don't create build/development files in source directories. Users should be able to specify where they want things to go / where input data is when they call the libraries. To test this framework, you should put an example in NaviGator for how to use it with NaviGator's specific classes/data/model.
* Any executables should be "thin", meaning they are essentialy just an argument parser which calls library functions.

**Comment by Grymessa** [comment]
This portion is ready to go. Some small quality of life improvements can still be made when it comes to building the docker image (like having it do so automatically through install script) but on further thought that would add a 14 gb docker image to the computer that might not be necessary and the command to generate a docker image is trivial.

File: perception/docker_tf/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,197 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(docker_tf)
+
+## Compile as C++11, supported in ROS Kinetic and newer
+# add_compile_options(-std=c++11)
+
+## Find catkin macros and libraries
+## if COMPONENTS list like find_package(catkin REQUIRED COMPONENTS xyz)
+## is used, also find other catkin packages
```
Comment (kev-the-dev) [inline perception/docker_tf/CMakeLists.txt]: remove boostrap comments in package.xml and CMakeLists.txt

File: perception/docker_tf/transfer_learning/data/fullcourse.pbtxt [hunk 2]
```diff
@@ -0,0 +1,52 @@
+item {
+    id: 1
+    name: 'None'
+    id: 2
+    name: 'target_large'
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/data/fullcourse.pbtxt]: Navigator specific items?

File: perception/docker_tf/transfer_learning/labelbox_data_processing/__init__.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+import pandas as pd
+import numpy as np
+import labelbox2pascal as lb2pa
+from PIL import Image
+from sklearn.model_selection import train_test_split
+
+dictionary = {}
+
+
+def split_data(image_dir='Images', ann_dir='Annotations', resize=False, size=[256, 150]):
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/labelbox_data_processing/__init__.py]: function comment, what do parameters do, what does it return.

File: perception/docker_tf/transfer_learning/labelbox_data_processing/json_files/navigator_initial.json [hunk 4]
```diff
@@ -0,0 +1 @@
..._chained1_2016-11-12-13-08-10_bag__stereo_right_image_raw10997932276976","Agreement":null,"Dataset Name":"navigator_pointgrey_wauburg_2016_11_12.json","Reviews":[],"View Label":"https://image-segmentation-v4.labelbox.com?project=cjm82v7349sed0780ftl7pawi&label=cjm854zi8a7h50780s268np3a"},{"ID":"cjm854y04h7y50728ibjtys9g","Labeled Data":"https://ci.mil.ufl.edu/labelme/datasets/navigator-datasets/navigator_pointgrey_wauburg_2016_11_12/2016-11-12_detect5_2016-11-12-14-55-52_bag__right_right_image_raw17204525191624.png","Label":{"circle_green":[{"geometry":[{"x":197,"y":238},{"x":197,"y":326},{"x":252,"y":326},{"x":252,"y":238}]}],"target_large":[{"geometry":[{"x":286,"y":162},{"x":286,"y":228},{"x":326,"y":228},{"x":326,"y":162}]}],"target_small":[{"geometry":[{"x":231,"y":180},{"x":231,"y":215},{"x":257,"y":215},{"x":257,"y":180}]}]},"Created By":"kma1660@gmail.com","Project Name":"navigator_computer_vision","Created At":"2018-09-18T19:58:07.000Z","Seconds to Label":14.591,"External ID":"2016-11-12_detect5_2016-11-12-14-55-52_bag__right_right_image_raw17204525191624","Agreement":null,"Dataset Name":"navigator_pointgrey_wauburg_2016_11_12.json","Reviews":[],"View Label":"https://image
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/labelbox_data_processing/json_files/navigator_initial.json]: navigator specific files

File: perception/docker_tf/transfer_learning/labelbox_data_processing/process.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
+from __future__ import division
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/labelbox_data_processing/process.py]: File should be renamed. "process" is ambiguous

File: perception/docker_tf/transfer_learning/labelbox_data_processing/generate_tfrecord.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
+"""
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/labelbox_data_processing/generate_tfrecord.py]: Is this file still used?

File: perception/docker_tf/transfer_learning/labelbox_data_processing/labelbox2pascal/__init__.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
+import os
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/labelbox_data_processing/labelbox2pascal/__init__.py]: We don't put code in `__init__`, we just use it to import from other files in a module

File: perception/docker_tf/README.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        return 5
+    elif row_label == 'Black Buoy':
+        return 6
+    else:
+        None
+```
+
+### labelmap.pbtxt
+
+Currently you have to manually edit this file as well, which is unfortunate. Ideally only one of these files should be manually edited. This contains all the labels and matching ID's you will be using for your identification. This is found in the data directory, which is likely one directory above the one you are currently in (you should see a data folder, a labelbox_data_processing folder, and a train.sh script. You'll also need to make a models folder if one is not present there and place a model in it that you want to retrain). An example of this file is the following:
```
Comment (kev-the-dev) [inline perception/docker_tf/README.md]: Is this still accurate? Be sure to go through README and make sure it _all_ reflects the latest changes

File: perception/docker_tf/transfer_learning/labelbox_data_processing/process.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        output_path = os.path.join(os.getcwd(), FLAGS.output_path)
+        print('Successfully created the TFRecords: {}'.format(output_path))
+
+    # Clean up
+    if FLAGS.cleanup:
+        try:
+            os.remove('test_labels.csv')
+            os.remove('train_labels.csv')
+            shutil.rmtree('Images/')
+            shutil.rmtree('Annotations/')
```
Comment (kev-the-dev) [inline perception/docker_tf/transfer_learning/labelbox_data_processing/process.py]: Aren't some of these configurable so shouldn't be hardcoded in cleanup?

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
Some initial comments regarding following conventions, documentation of library functions with no comments, and redundant files.

File: perception/mil_mlp/README.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+This process starts with labelling images on the service Labelbox, which as it currently stands is the standard for image labeling.
+Once the data is labelled you can then download this image set in a variety of formats. This pipeline currently supports two:
+
+1. JSON
+2. PASCAL VOC
+
+Of these two, it has been most extensively tested with JSON files and thus we recommend downloading the dataset in this format.
+JSON files for MIL projects are found in their respective repositories (i.e. the NaviGator and SubjuGator repositories).
+
+#### A. process.py
```
Comment (kev-the-dev) [inline perception/mil_mlp/README.md]: now process_data (it is still referred to as process.py in a few places)

File: perception/mil_mlp/README.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+5. Configurable CSV input (Currently it still assumes you are handing it a json or PASCAL so it will error out if
+those are not provided.)
+6. Add python dependencies to the install script.
+
+## Quick Start Guide
+
+1. Process the Data:
+
+```bash
+rosrun mil_mlp process_data.py --json_path='json_files/JSON_NAME.json --labelmap_path='../docker_tf/transfer_learning/data/LABELMAPNAME.pbtxt
```
Comment (kev-the-dev) [inline perception/mil_mlp/README.md]: Still would prefer a less ambiguous name for this script. Why not `convert_labels_to_tfrecord`?

File: perception/mil_mlp/README.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+```
+2. Run the docker image
+```bash
+rosrun mil_mlp pipeline.sh
```
Comment (kev-the-dev) [inline perception/mil_mlp/README.md:26]: also here it would be nice to have a descriptive name, like `start_tensorflow_training_container`

File: perception/mil_mlp/README.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+This pipeline is built upon Docker Images for easy deployment and updating. Currently, we build and update off of 3 distinct Docker Images. The first image is the CUDA 10.0 docker image distributed by NVIDIA. This simply installs CUDA drivers into the image and allows us to communicate with the card. The second image we  run is a custom tensorflow 9.0 image, compiled using the dockerfiles made availble to us from Tensorflow. We had to configure the compilation to target CUDA 10.0 instead of CUDA 9.0, which is the default. CUDA 10.0 is not officially supported by TF as of yet, however it is the only version of CUDA our graphics card supports. The final image is the docker image that loads our custom scripts and the Models directory of Tensorflow. Eventually, these will all be merged into a single docker image, however this allows us to update CUDA, Tensorflow, and our scripts separately, without having to recompile everything each time.
+
+### II. The Data: ldp
```
Comment (kev-the-dev) [inline perception/mil_mlp/README.md]: Mention what ldp stands for, took me a while to figure it out

File: perception/mil_mlp/README.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
+# Transfer Learning Object Detection - A Short Tutorial
```
Comment (kev-the-dev) [inline perception/mil_mlp/README.md]: Before jumping into usage, nice to provide a description of what this package is / does, you can even just copy this from package.xml

File: perception/mil_mlp/README.md (hunk 12 above)
Comment (kev-the-dev) [inline perception/mil_mlp/README.md:26]: Looks like you need to be in a directory that contains a model already downloaded before running this, please note that here

File: perception/mil_mlp/README.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+# @TODO
+1. Configurable image download
+2. Add docker build to the install script. Until then docker build must be done manually.
+3. Have datasets stored on internal server so it can be downloaded.
+4. Configurable resize (Low importance due to no significant performance gain)
+5. Configurable CSV input (Currently it still assumes you are handing it a json or PASCAL so it will error out if
+those are not provided.)
+6. Add python dependencies to the install script.
```
Comment (kev-the-dev) [inline perception/mil_mlp/README.md:10]: This should happen before merging. Developers should be able to run the code in our repo after running install script

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
Some minor (quick to fix!) final concerns, otherwise this is looking good. Long term, functionality should be added to support different models (instead of one hardcoded one), but that can be future work

**Comment by kev-the-dev** [comment]
Also there seems to be a build error currently (see CI), and I imagine some python to in need of autoformating (you can use `mcfmt` to verify)

File: perception/mil_mlp/README.md (hunk 15 above)
Comment (Grymessa) [reply perception/mil_mlp/README.md:10]: Only caveat is that right now we have to build a custom tensorflow image, which implies that the user would need to download the full tensorflow repo and run their custom script to generate a docker image. Unfortunately these images are going to be tied to the specific GPU you are running.

**Review by kev-the-dev — APPROVED** [review]
Tentative approval pending a passing build on CI

File: perception/mil_mlp/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    docker_tf
+  CATKIN_DEPENDS
+    rospy
+  DEPENDS
+)
+
+
+catkin_python_setup()
+
+include_directories(
```
Comment (kev-the-dev) [inline perception/mil_mlp/CMakeLists.txt]: This is failing the build and I believe is unnecessary... is there any C++ for this project? Python does not use include_directories

**Review by kev-the-dev — CHANGES_REQUESTED** [review]

**Comment by kev-the-dev** [comment]
Reminder that this PR is almost mergable just need to get CI happy

**Comment by Grymessa** [comment]
> Reminder that this PR is almost mergable just need to get CI happy

It is now happy
